queue.go

```diff
@@ -95,47 +97,113 @@ func (q *MemTxQueue) Push(tx Transaction) {
 	q.txs = append(q.txs, tx)
 }
 
+type TxValidator func(Transaction) bool
+
+// TxQueueManager manages transaction queue. It receive transaction from client
+// and TxQueueManager have its own policy to propose contribution to honeybadger
 type TxQueueManager interface {
 	AddTransaction(tx Transaction) error
 }
 
 type DefaultTxQueueManager struct {
 	txQueue TxQueue
 	hb      HoneyBadger
+
+	stopFlag int32
+
+	batchSize int
+	blockSize int
+
+	closeChan chan struct{}
+
+	tryInterval time.Duration
+
+	txValidator TxValidator
 }
 
-func NewDefaultTxQueueManager(txQueue TxQueue, hb HoneyBadger) *DefaultTxQueueManager {
-	return &DefaultTxQueueManager{
-		txQueue: txQueue,
-		hb:      hb,
+func NewDefaultTxQueueManager(
+	txQueue TxQueue,
+	hb HoneyBadger,
+	batchSize int,
+	blockSize int,
+
+	// tryInterval is time interval to try create contribution
+	// then propose to honeybadger component
+	tryInterval time.Duration,
+
+	txVerifier TxValidator,
+) *DefaultTxQueueManager {
+	m := &DefaultTxQueueManager{
+		txQueue:   txQueue,
+		hb:        hb,
+		batchSize: batchSize,
+		blockSize: blockSize,
+
+		closeChan:   make(chan struct{}),
+		tryInterval: tryInterval,
+		txValidator: txVerifier,
 	}
-}
 
-func (m *DefaultTxQueueManager) ReceiveBatch(batch Batch) {
-	// send consensused batch to application
+	go m.runContributionProposeRoutine()
 
-	// follow default policy then create batch
+	return m
 }
 
 func (m *DefaultTxQueueManager) AddTransaction(tx Transaction) error {
-	// validate transaction
-
+	if !m.txValidator(tx) {
+		return errors.New(fmt.Sprintf("error invalid transaction: %v", tx))
+	}
 	m.txQueue.Push(tx)
-
 	return nil
 }
 
+func (m *DefaultTxQueueManager) Close() {
+	if first := atomic.CompareAndSwapInt32(&m.stopFlag, int32(0), int32(1)); !first {
+		return
+	}
+	m.closeChan <- struct{}{}
+	<-m.closeChan
+}
+
+func (m *DefaultTxQueueManager) toDie() bool {
+	return atomic.LoadInt32(&(m.stopFlag)) == int32(1)
+}
+
+// runContributionProposeRoutine tries to propose contribution every its "tryInterval"
+// And if honeybadger is on consensus, it waits
+func (m *DefaultTxQueueManager) runContributionProposeRoutine() {
+	for !m.toDie() {
+		if !m.hb.OnConsensus() {
+			m.tryPropose()
+		}
+		time.Sleep(m.tryInterval)
+	}
+}
+
+// tryPropose create contribution and send it to honeybadger only when
+// transaction queue size is larger than batch size
+func (m *DefaultTxQueueManager) tryPropose() error {
+	if m.txQueue.Len() < int(m.batchSize) {
+		return nil
+	}
+	contribution, err := m.createContribution()
+	if err != nil {
+		return err
+	}
+	return m.hb.HandleContribution(contribution)
+}
+
 // Create batch polling random transaction in queue
-func (m *DefaultTxQueueManager) createBatch(n, b int) (Batch, error) {
-	candidate, err := m.loadCandidateTx(min(b, m.txQueue.Len()))
+func (m *DefaultTxQueueManager) createContribution() (Contribution, error) {
+	candidate, err := m.loadCandidateTx(min(m.blockSize, m.txQueue.Len()))
```

Comment:
what's purpose of `min(m.blockSize, m.txQueue.Len())`?

Reply:
in this step, we pick up candidates as much as block size and this is max value. if transaction queue has not that size of transaction, pick all of the transaction from the queue